distributed/scheduler.py

```diff
@@ -1590,6 +1596,33 @@ def get_nbytes(self, stream=None, keys=None, summary=True):
 
             return result
 
+    def publish_dataset(self, stream=None, keys=None, data=None, name=None,
+                        client=None):
+        self.client_wants_keys(keys, 'published-%s' % name)
+        self.published_data[name] = {'data': data, 'keys': keys,
+                                     'clients': {client}}
```

Comment:
The executor may not know about what keys are available if it hasn't synchronized with the scheduler recently.  Even if we do try to synchronize with the scheduler then we open ourselves up to race conditions.  I recommend checking this on the scheduler side, where everything is sequential.